Updated package compatibility.

JorgenVatle opened the pull request:
The `\Illuminate\Support\Str::after()` method used in the `indisposable` validator was added in version `5.4` of `illuminate/support`, which makes this package incompatible with earlier versions.

This pull request updates the `illuminate/support` version requirement in `composer.json` to `5.4.*` and adds a compatibility notice in the `README.md`. :smile:

JorgenVatle committed:
Added package compatibility notice to readme.

Laravel's `\Illuminate\Support\Str::after()` method was added in version `5.4`, which makes this package incompatible with earlier versions. :smile:

JorgenVatle committed:
Corrected `compatibility` typo in readme.

JorgenVatle committed:
Updated `Illuminate/Support` version requirement.

Propaganistas commented:
If `Str::after()` is the only culprit forcing to 5.4+ I'd rather opt for dropping the helper instead. If you're willing to change your PR accordingly and clean it up I'd be happy to merge. Thanks!

JorgenVatle committed:
Added PHPUnit testing framework to dev dependencies.

JorgenVatle committed:
Added PHPUnit base configuration.

JorgenVatle committed:
Added `Illuminate\Cache` to dependencies.

JorgenVatle committed:
Added cache fetcher method.

JorgenVatle committed:
Added cache storage method.

JorgenVatle committed:
Made DisposableEmail cache storage method static.

JorgenVatle committed:
Relying on the DisposableEmail caching class for cache storage.

JorgenVatle committed:
Updated scope of the $sourceUrl attribute in DisposableEmail Cache.

JorgenVatle committed:
Added DisposableEmail cache fetch method.

JorgenVatle committed:
Added test for the cache source fetching method. (`fetchSource()`)

JorgenVatle committed:
Added test for fetching domain sources from the locally stored domain json file.

JorgenVatle committed:
Added local domain source cache fetching method.

JorgenVatle committed:
Renamed the Cache remote domain source fetching method to better comply with the current naming scheme.

JorgenVatle committed:
Updated PHPDoc to expect array in the Cache `store()` method.

JorgenVatle committed:
Added Cache data storage test.

JorgenVatle committed:
Added Orchestra's Laravel Testbench to dev-dependencies.

JorgenVatle committed:
Using Orchestra's Laravel Testbench TestCase component over the PHPUnit TestCase component.

JorgenVatle committed:
Added Laravel Framework version requirement.

JorgenVatle committed:
Removed version requirement from readme.

JorgenVatle committed:
Added DisposableEmail Cache update method.

JorgenVatle committed:
Corrected domain format for the sake of consistency.

JorgenVatle committed:
Added DisposableEmail Cache update test.

JorgenVatle committed:
Added DisposableEmail Cache source string decoding method.

JorgenVatle committed:
Automatically decoding data source before pushing it to the cache in DisposableEmail Cache store method.

JorgenVatle committed:
Removed unnecessary json_decoding calls in DisposableEmail Cache `update()` method.

JorgenVatle committed:
Asserting that the DisposableEmail Cache `store()` method is able to handle JSON strings.

JorgenVatle committed:
Updated scope of the DisposableEmail Cache `decodeSource()` method.

JorgenVatle committed:
Corrected DisposableEmailCacheTest class name.

JorgenVatle committed:
Added DisposableEmail cache flush method.

JorgenVatle committed:
Added test for DisposableEmail Cache `fetchOrUpdate()` method - ensuring the method properly renews and fetches the cache.

JorgenVatle committed:
Added DisposableEmail Cache `fetchOrUpdate()` method.

JorgenVatle committed:
Updated Indisposable class name.

JorgenVatle committed:
Added test for disposable email domain detection.

JorgenVatle committed:
Ported over core IndisposableValidation functionality to dedicated class for better testability and reusability.

JorgenVatle committed:
Removed unnecessary DisposableEmail Cache import.

JorgenVatle committed:
Added Indisposable import and updated `isDisposable()` method name.

JorgenVatle committed:
Updated test method name for non-disposable test to better reflect what we're actually testing for.

JorgenVatle committed:
Added DisposableEmail test for disposable emails.

JorgenVatle committed:
Added Cache tests setup method.

JorgenVatle committed:
Removed unused Illuminate Cache import from DisposableEmailCache tests.

JorgenVatle committed:
Updated TestCase import.

JorgenVatle committed:
Refreshing cache on DisposableEmail test setup.

JorgenVatle committed:
Corrected return type of invalidly formatted emails in `isDisposable()` method.

JorgenVatle committed:
Added Indisposable facade.

JorgenVatle committed:
Registered Indisposable binding.

JorgenVatle committed:
Updated Indisposable Facade class name.

JorgenVatle committed:
Added package TestCase.

JorgenVatle committed:
Updated naming scheme for Indisposable facade.

JorgenVatle committed:
Updated Indisposable Facade class name.

JorgenVatle committed:
Using absolute path for Orchestra Testbench.

JorgenVatle committed:
Added Orchestra Testbench service providers.

JorgenVatle committed:
Corrected facade alias for Indisposable.

JorgenVatle committed:
Converted `isDisposable()` to a dynamic method to ensure the constructor is called.

JorgenVatle committed:
Using Indisposable facade in validation method.

JorgenVatle committed:
Using Indisposable facade in Indisposable tests.

JorgenVatle committed:
Added DisposableEmail Test namespaces.

JorgenVatle committed:
Added DisposableEmail Tests autoloader.

JorgenVatle committed:
Updated naming scheme for validation tests.

JorgenVatle committed:
Added tests for the Validator's `validate()` method.

JorgenVatle committed:
Using local version of Illuminate's Str::after() helper.

JorgenVatle committed:
Corrected Validation tests namespacing.

JorgenVatle committed:
Corrected improperly named class name in IndisposaableValidationTest.

JorgenVatle committed:
Corrected improperly named IndisposableValidationTest class name.

JorgenVatle committed:
Merge remote-tracking branch 'origin/patch-1' into patch-1

JorgenVatle commented:
Done 👍 

While I was at it, I did some basic refactoring and added test coverage along with a facade.

I may have gone a bit overboard on this one. If you're not a fan of the changes I can of course go ahead and revert back to a quick and simple fix as originally requested 😅

JorgenVatle committed:
Added Indisposable default error message.

JorgenVatle committed:
Added default error message to `indisposable` validator extension.

JorgenVatle committed:
Made language file entry optional as we now have a default error message in English.

JorgenVatle commented:
[+] Default validation error message.

Drops the requirement for the user needing to add an entry to their languages file for user-friendly validation errors.

Propaganistas commented:
Wow, you're definitely a big fan of the package! Great work. There's still room for improvement. Some guidelines:
* Is a separate `Cache` class really necessary? I'd merge it in the new `Indisposable` class.
* Make use of the Container for constructing `Indisposable`. E.g. inject the cache instead of referencing it through a facade, ...
* Since fetching the domains list is now done in `Indisposable`, move the `isUsefulJson()` method over there and make the fetching methods use it as well. The command should just execute an external fetch and provide info likewise. It should not be concerned about the internals of that fetch.
* Perhaps simplify all those fetching methods? Are they really necessary? Would any developer ever use them directly? The command exposes enough in my opinion, namely updating the list.
* Remove staticness. `Indisposable` is a singleton.

Again, if you're willing to clean this up I'd be happy to merge :-)

JorgenVatle commented:
Thanks for the input! 😄 

I'll go ahead and apply your guidelines shortly. 👍 

Quick question - how would you like to handle non-useful JSON in the Indisposable class in conjunction with the cache updater command? Should I throw an exception and output the exception message to the console as an error? Or do you have something else in mind?

Propaganistas commented:
Throwing an exception and catching it in the command seems the way to go.

JorgenVatle committed:
Added pre-parsed and cached version of the local array of disposable domains.

JorgenVatle committed:
Added remote disposable domains cache method.

JorgenVatle committed:
Added domains method with fallback.

JorgenVatle committed:
Updated Indisposable domain and helper method scopes, opting for `protected` over `public`.

JorgenVatle committed:
Removed staticness of the Indisposable `domains` attribute.

JorgenVatle committed:
Using domains from self in Indisposable `domains` attribute.

JorgenVatle committed:
Opting for assigning the Indisposable `domains` attribute through the constructor instead of a dedicated method.

JorgenVatle committed:
Added cache flush method to Indisposable.

JorgenVatle committed:
Logging remote domain fetch exceptions.

JorgenVatle committed:
Updated Indisposable method order formatting.

JorgenVatle committed:
Relying on Indisposable class for flushing the domain cache.

JorgenVatle committed:
Using Indisposable facade for flushing the cache on IndisposableValidation test setup.

JorgenVatle committed:
Registering Indisposable binding as a singleton.

JorgenVatle committed:
Updated scope of Indisposable `remoteDomains` method.

JorgenVatle committed:
Asserting that the Indisposable domain cache can be flushed.

JorgenVatle committed:
Added IDE-helping PHPDoc to Indisposable facade.

JorgenVatle committed:
Added `$email` parameter to Indisposable Facade's isDisposable IDE helper PHPDoc.

JorgenVatle committed:
Asserting that the Indisposable `remoteDomains()` method builds a cache.

JorgenVatle committed:
Added PHPDoc to indisposable email cache key property.

JorgenVatle committed:
Relying on Indisposable class for updating the disposable domain cache.

JorgenVatle committed:
Removed unused `isUsefulJson()` method.

JorgenVatle committed:
Removed unused `Cache` class.

JorgenVatle committed:
Removed Indisposable `Cache` unit tests.

JorgenVatle committed:
Asserting that the disposable domain cache command does indeed update the domain cache.

JorgenVatle committed:
Updated scope of the local domains method to allow for external use and seamless testability.

JorgenVatle committed:
Added local domains to Indisposable Facade's IDE helping PHPDoc.

JorgenVatle committed:
Asserting that the Indisposable remoteDomains array can be fetched.

JorgenVatle committed:
Asserting that the Indisposable local domains can be loaded.

JorgenVatle committed:
Updated IndisposableTest test method formatting.

JorgenVatle committed:
Removed unused imports from IndisposableTest class.

JorgenVatle committed:
Asserting that our validation methods are usable through the Laravel Validator facade.

JorgenVatle committed:
Handling singular domain caching in Disposable domain caching command info output.

JorgenVatle commented:
I believe that should be it. Let me know if there's anything else you'd like me to change. 👍 

I left the local and remote domain lists public as they may prove useful to package users and makes testing a tad bit less cumbersome.

**Summary of changes:**
- Merged the `Cache` class' functionality with `Indisposable`.
- Moved `CacheDisposableDomainsCommand`'s `isUsefulJson()` method to `Indisposable`.
- Removed staticness of the `Indisposable` class.
- Storing caches for local and remote domains indefinitely, until renewed by the cache disposable domains command.
- Added tests asserting that the `Indisposable` validator and cache command has extended the Laravel and works as intended.

Propaganistas commented on composer.json:
```diff
@@ -19,13 +19,12 @@
-        "illuminate/support": ">=5.0,<5.7",
-        "illuminate/console": ">=5.0,<5.7",
-        "illuminate/validation": ">=5.0,<5.7"
+        "laravel/framework": ">=5.0.0"
```
What's the reasoning behind this change?

JorgenVatle replied:
The `orchestra/testbench` package used to run tests in a Laravel environment appeared to take issue with the requirements.

After double checking, I can't seem to replicate the issue. I'll go ahead and update the requirements accordingly. Thanks for bringing that up.

Propaganistas commented on src/Console/CacheDisposableDomainsCommand.php:
```diff
@@ -20,72 +21,28 @@ class CacheDisposableDomainsCommand extends Command
 
-            return;
-        }
+            Indisposable::flushCache();
```
Consider the following scenario:
1) Cache gets flushed
2) Remote domains are fetched, but fail
3) Even though there was a valid domains list in cache, you just flushed them away and are now forced to use the list shipped in the package instead.

-> Store remote domains on disk and use that one as a first fallback. Use the shipped package domains only as a final fallback if remote domains were never fetched.

JorgenVatle replied:
Oh, good catch!

The issue has been addressed with the new update, along with a test case replicating the scenario. 😃

Propaganistas commented on src/DisposableEmailServiceProvider.php:
```diff
@@ -31,6 +36,8 @@ public function boot()
     {
-        //
+        $this->app->singleton('indisposable', function() {
+            return new Indisposable();
```
Supply `$app['cache']` here instead of relying on Facades in the class itself

Propaganistas commented on src/Validation/Indisposable.php:
```diff
@@ -2,52 +2,120 @@
-     * @var array
+     * @var string
+     */
+    protected $sourceUrl = 'https://rawgit.com/andreis/disposable-email-domains/master/domains.json';
```
Make it `public static` or implement a setter so the source URL can be changed when desired.

Propaganistas commented on src/Validation/Indisposable.php:
```diff
@@ -2,52 +2,120 @@
+     * @var string
      */
-    protected static $domains = [];
+    protected $cacheKey = 'laravel-disposable-email.cache';
```
Change to `disposable-email.domains`

Propaganistas commented on src/Validation/Indisposable.php:
```diff
@@ -2,52 +2,120 @@
+        try {
+            $this->domains = $this->remoteDomains();
+        } catch (Exception $exception) {
+            Log::warning($exception->getMessage());
```
No need to implement logging in a package. It's up to the developer to decide if logging is desired in the ExceptionHandleR.

JorgenVatle replied:
Got it! Thanks for the explanation. 👍

Propaganistas commented on src/Validation/Indisposable.php:
```diff
@@ -2,52 +2,120 @@
 namespace Propaganistas\LaravelDisposableEmail\Validation;
 
-use Illuminate\Support\Str;
+use Illuminate\Support\Facades\Cache as FrameworkCache;
```
See ServiceProvider -> no facades.

Propaganistas commented on src/Validation/IndisposableValidation.php:
```diff
@@ -0,0 +1,30 @@
+     *
+     * @var string
+     */
+    static $errorMessage = 'Disposable email addresses are not allowed.';
```
Explicit property visibility. Better make it protected as well.

JorgenVatle replied:
The `$errorMessage` property is used by the service provider for assigning a default validation error message.

> **src/DisposableEmailServiceProvider.php:**
```php
 $this->app['validator']->extend(
    'indisposable',
    IndisposableValidation::class . '@validate',
    IndisposableValidation::$errorMessage
);
```

Would you like me to instead remove the property and assign the error message inline with the validator extension?

Propaganistas commented on tests/Validation/IndisposableTest.php:
```diff
@@ -0,0 +1,76 @@
+     *
+     * @var string
+     */
+    private $cacheKey = 'laravel-disposable-email.cache';
```
Change accordingly

Propaganistas commented on tests/Validation/IndisposableTest.php:
```diff
@@ -0,0 +1,76 @@
+
+namespace Propaganistas\LaravelDisposableEmail\Tests\Validation;
+
+use Illuminate\Support\Facades\Cache as FrameworkCache;
```
No need to alias the cache

Propaganistas reviewed (changes requested):
Thanks for your effort. Here's some final feedback.

JorgenVatle committed:
Replaced `laravel/framework` dependency with `illuminate` packages.

JorgenVatle committed:
Relying on `Illuminate\Cache` dependency injection for caching in `Indisposable` class.

JorgenVatle committed:
Updated disposable email cache key.

JorgenVatle committed:
Removed Indisposable warning logger.

JorgenVatle committed:
Added Indisposable cache key getter.

JorgenVatle committed:
Added Indisposable remote URL setter.

JorgenVatle committed:
Updated remote URI attribute to better conform to the current naming scheme.

JorgenVatle committed:
Added IDE helper for Indisposable `getCacheKey()` method.

JorgenVatle committed:
Added IDE helper for Indisposable `setRemoteUrl` method.

JorgenVatle committed:
Returning updated value on Indisposable `setRemoteUrl()` call.

JorgenVatle committed:
Fetching cache key from Indisposable facade on IndisposableTest setup.

JorgenVatle committed:
Specified scope of the `IndisposableValidation` `$errorMessage` property.

JorgenVatle committed:
Added remote domain cache override method.

JorgenVatle committed:
Added IDE helper for Indisposable `setRemoteDomainsCache()` method.

JorgenVatle committed:
Removed unnecessary alias of `Cache` facade.

JorgenVatle committed:
Using Cache `forever()` method for overriding the domains cache.

JorgenVatle committed:
Asserting that the Indisposable cache override method overrides the domain cache.

JorgenVatle committed:
Using domain name in IndisposableTest cache override test instead of email address for the sake of consistency with expected data.

JorgenVatle committed:
Re-caches flushed domains on cache update exception.

JorgenVatle committed:
Added missing parameter definition to Indisposable Facade's `setRemoteUrl()` IDE helper.

JorgenVatle committed:
Asserting that the domain cache is rebuilt if an exception is encountered whilst updating the remote domain cache in CacheDisposableDomainsCommand.

JorgenVatle committed:
Extended IndisposableTest domain cache re-instantiation assertion.

JorgenVatle commented:
I believe these changes should have addressed the issues you pointed out. 🙂 

Let me know if I missed something.

Propaganistas commented:
Could you rebase the PR on the `dev-refactor` branch please? I'm going to make a few additional tweaks myself before merging it into `master`.

JorgenVatle commented:
There you go. 👍

JorgenVatle committed:
No longer calling Indisposable's non-static `stringAfter()` method statically from `isDisposable()`

JorgenVatle committed:
Moved Indisposable's `stringAfter()` method functionality inline with `isDisposable()`'s Regex call.

JorgenVatle committed:
Improved readability of Indisposable's `isDisposable` regex call.

JorgenVatle committed:
Asserting that invalid email addresses are discarded by Indisposable's `isDisposable()` method.

JorgenVatle commented:
Went ahead and replaced `Indisposable`'s `stringAfter()` method with a readable inline solution in `isDisposable()`.

Thought there wasn't much need to keep that method around for that one purpose.

JorgenVatle committed:
Extended test cases for `isDisposable` discards.

JorgenVatle committed:
Added limit to `isDisposable()`'s `explode()` call.
